src/OAuth/OAuth2/Service/BattleNetCN.php

```diff
@@ -0,0 +1,18 @@
+<?php
+namespace OAuth\OAuth2\Service;
+
+use OAuth\Common\Http\Uri\Uri;
+use OAuth\Common\Consumer\Credentials;
+use OAuth\Common\Http\Client\ClientInterface;
+use OAuth\Common\Storage\TokenStorageInterface;
+use OAuth\Common\Http\Uri\UriInterface;
+
+class BattleNetCN extends BattleNetBase
+{
+    public function __construct(Credentials $credentials, ClientInterface $httpClient, TokenStorageInterface $storage, $scopes = array(), UriInterface $baseApiUri = null)
```

Comment:
I understand what your trying to achieve, using the service factory to generate a service that is already pointed to the correct region. We have this ability already through the ServiceFactory:

```php
$battleNet = $serviceFactory->createService(
    'BattleNet',
    $credentials,
    $storage,
    []
    new Uri(BattleNet::REGION_CN)
);
```

Then you may enforce requirements that this must be provided quite easily by validating that `$baseApiUri` is not `null` in your constructor.

Or if the developer wanted this level of abstraction they could create their own BattleNetServiceFactory.

The region is just a configuration parameter of the service, much like the environment of some other services carry their own URL and must be set in the same way. Its not appropriate to create different services that are all really the same service with slightly different configuration.